README.txt
= Rubyforge

* http://codeforpeople.rubyforge.org/rubyforge/
* http://rubyforge.org/projects/codeforpeople/

== Description

A script which automates a limited set of rubyforge operations.

* Run 'rubyforge help' for complete usage.
* Setup: For first time users AND upgrades to 0.4.0:
  * rubyforge setup
  * edit ~/.rubyforge/user-config.yml
  * rubyforge config
* Don't forget to login!  logging in will store a cookie in your
  .rubyforge directory which expires after a time.  always run the
  login command before any operation that requires authentication,
  such as uploading a package.

== Synopsis

  rubyforge [options]* mode [mode_args]*



History.txt
=== 2.0.2 / 2009-10-05

* Switched JSON gem dependency over to json_pure.

=== 2.0.1 / 2009-10-02

* Added a missing dependency on JSON gem to the spec.

=== 2.0.0 / 2009-09-21

* Modified to use RubyForge REST API rather than scraping HTML.

=== 1.0.5 / 2009-09-15

* Added experimental script to sync RF trackers with release names.
* Fixed that damned processor_id bug. (hinegardner)
* The century usually doesn't change across runs. Refactor.
* Fix use of Time.utc if args.size < 10, tho seems unnecessary.

=== 1.0.4 / 2009-07-21

* Uses the passed in proxy, if provided. (Thanks sdabet, RF #24071).
* Update group_id pattern for scraping project configs you are only a member of.
* Update regexp used to validate login page to be less restrictive.
* Fixed --help to not require an argument.
* add --force flag for login command to ignore previous cookie. (or use logout)

=== 1.0.3 / 2009-02-26

* Fixed nil error in our Net::HTTP patches.
* Removed password from warning if the login possibly failed.

=== 1.0.2 / 2009-01-05

* All webby commands now login automatically.
* Login now no-ops if it already has a session cookie.
* Added logout command.
* Much more of the config is self-repairing, but still not bulletproof yet.

=== 1.0.1 / 2008-10-22

* Fixed multipart form upload so it isn't url escaping the data. DOH.
  * Affects release_notes and release_changes, but never reported for 5 months.

=== 1.0.0 / 2008-05-20

* Removed HTTPAccess2, thanks to Aaron Patterson. Even tho he's whiny.
* Changed initialize/configure to make testing scream. 100x faster.

=== 0.4.5 / 2008-03-11

* Update for Ruby 1.9.0.
* Updated History, Rakefile, and Readme for new hoe abilities.
* Added config backup/restore rake tasks (for testing).

=== 0.4.4 / 2007-08-13

* New type_id values will merge with extant data. (self-repairing data is Good)
* Scrape processor_ids, merging in with extant data.
* Default to "Other" if a file's type is unrecognized.

=== 0.4.3 / 2007-07-23

* Set mode on .rubyforge directory to 700.
* Fix fetching of user id when user has no releases.

=== 0.4.2 / 2007-05-21

* Fix for windoze users (spaces in path).
* Added check for extant release.
* Added default hash for first-time releases.

=== 0.4.1 / 2007-03-08

* Verify that login succeeded and warn against if not (prolly should raise).
* Print a friendly error if you have the wrong package id.
* Handle upload error in add_release a bit better.

=== 0.4.0 / 2007-01-09

* config.yml split and moved to user-config.yml (up to the user to do).
* auto-config.yml now generated via config command.
* @config renamed to @userconfig.
* @config["rubyforge"] moved to @autoconfig.
* Added save_autoconfig.
* Pulled scrape_project from scrape_config.
* scrape_config no longer takes a user param. Use opts to specify.
* scrape_project, add_project, add/remove_release now save automatically.

=== 0.3.2 / 2006-11-29

* Fixed file uploads for windows.
* Correctly scrape releases with funky characters.

=== 0.3.1 / 2006-10-24

* Added SSL login.
* Added yet more debugging output if $DEBUG.

=== 0.3.0 / 2006-09-30

* Added more debugging output if $DEBUG
* Added news posting.
* Added multiple file release to add_release (uses add_file for extras).
* add_release now returns release_id
* Fixed config scraper to include '-' in names.

=== 0.2.1 / 2006-09-14

* Gemspec was too loose about packaging. Now using manifest.

=== 0.2.0 / 2006-09-13

* Split original script into script and library.
* Added tests for library.
* Refactored heavily.
* Added "config" command to scrape group/project/release ids from rubyforge.
* Added "names" command to help pick groups and projects.
* Added "add_file" command to add a file to an existing release.


test/test_rubyforge.rb
require 'test/unit' unless defined? $ZENTEST and $ZENTEST

$TESTING = true
require 'rubyforge'
require 'tmpdir'

class RubyForge
  attr_writer :client

  alias :old_save_autoconfig :save_autoconfig
  def save_autoconfig
    # raise "not during test"
  end
end

class RubyForge::FakeClient
  def form; end
  def save_cookie_store(*args) end

  def post_content(*args)
    FakeRubyForge::HTML
  end

  def get_content(*args)
    URI::HTTP.data.join("\n")
  end
end

class FakeRubyForge < RubyForge
  HTML = "blah blah <form action=\"/frs/admin/editrelease.php?group_id=440&release_id=6948&package_id=491\" method=\"post\"> blah blah"

  attr_accessor :page, :form, :extheader, :requests, :scrape
  def run(page, form, extheader={})
    @page, @form, @extheader = page, form, extheader
    @requests ||= []
    @requests << { :url => page, :form => form, :headers => extheader }
    HTML
  end

  def scrape_project(proj)
    @scrape ||= []
    @scrape << proj
  end
end

# TODO: remove this and make rubyforge use Client exclusively
class URI::HTTP
  def self.data
    @data ||= []
  end

  def read
    self.class.data.shift or raise "no more data"
  end
end

class TestRubyForge < Test::Unit::TestCase
  def setup
    srand(0)
    util_new FakeRubyForge
  end

  def teardown
    File.unlink @cookie_path if defined? @cookie_path
    #     if defined? @old_autoconfig then
    #       @rubyforge.autoconfig.replace @old_autoconfig
    #       @rubyforge.save_autoconfig
    #     end
  end

  def test_initialize_bad
    @cookie_path = File.join(Dir.tmpdir, "cookie.#{$$}.dat")
    File.open(@cookie_path, 'w') {} # touch

    user_data = {
      "uri"        => "http://rubyforge.org",
      "is_private" => false,
      "cookie_jar" => @cookie_path,
      "username"   => "username",
      "password"   => "password"
    }

    assert_raise RuntimeError do
      rf = RubyForge.new user_data
      rf.configure "username" => nil
    end
    assert_raise RuntimeError do
      rf = RubyForge.new user_data
      rf.configure "password" => nil
    end
    assert_raise RuntimeError do
      rf = RubyForge.new user_data
      rf.configure "cookie_jar" => nil
    end
  end

  def test_setup
    # TODO raise NotImplementedError, 'Need to write test_setup'
  end

  def test_login
    u, p = 'fooby', 's3kr3t'
    @rubyforge.userconfig['username'] = u
    @rubyforge.userconfig['password'] = p
    @rubyforge.login

    util_run('https://rubyforge.org/account/login.php',
             'form_pw' => p,
             'form_loginname' => u,
             'return_to' => '',
             'login' => 'Login')
  end

  def test_create_package
    @rubyforge.create_package(42, 'woot_pkg')

    util_run('/frs/admin/index.php',
             "submit" => "Create This Package",
             "group_id" => 42,
             "is_public" => 1,
             "package_name" => "woot_pkg",
             "func" => "add_package")
  end

  def test_delete_package
    @rubyforge.delete_package(42, 666)
    util_delete_package
  end

  def test_delete_package_package_name
    @rubyforge.delete_package(42, "woot_pkg")
    util_delete_package
  end

  def test_delete_package_undefined_package_name
    assert_raise RuntimeError do
      @rubyforge.delete_package(42, "blah")
    end
  end

  def test_delete_package_group_name
    @rubyforge.delete_package("seattlerb", 666)
    util_delete_package
  end

  def test_delete_package_undefined_group_name
    assert_raise RuntimeError do
      @rubyforge.delete_package("blah", 666)
    end
  end

  def test_add_file
    @rubyforge.autoconfig["package_ids"]["ringy_dingy"] = 314
    @rubyforge.autoconfig["release_ids"]["ringy_dingy"] ||= {}
    @rubyforge.autoconfig["release_ids"]["ringy_dingy"]["1.2.3"] = 43

    @rubyforge.add_file('seattlerb', 'ringy_dingy', '1.2.3', __FILE__)

    util_run('/frs/admin/editrelease.php?group_id=42&release_id=43&package_id=314',
             { "step2" => 1,
               "type_id" => 9999,
               "processor_id" => 8000,
               "submit" => "Add This File"},
             {"content-type"=> "multipart/form-data; boundary=___00__03__03__39__09__19__21__36___"})
  end

  def test_add_release
    @rubyforge.add_release(42, 666, '1.2.3', __FILE__)
    util_add_release
  end

  def test_add_release_multiple
    @rubyforge.add_release(42, 666, '1.2.3', __FILE__, __FILE__) # dunno if that'll work
    add_release = ({ :url=>"/frs/admin/qrs.php",
                     :form=>{"processor_id"=>8000,
                       "submit"=>"Release File",
                       "preformatted"=>0,
                       "release_changes"=>nil,
                       "type_id"=>9999,
                       "group_id"=>42,
                       "release_name"=>"1.2.3",
                       "release_notes"=>nil,
                       "package_id"=>666,
                       "release_date"=>"today"},
                     :headers=> {"content-type" => "multipart/form-data; boundary=___00__03__03__39__09__19__21__36___"}})
    add_file = ({ :url => '/frs/admin/editrelease.php?group_id=42&release_id=6948&package_id=666',
                  :form => { "step2" => 1,
                    "type_id" => 9999,
                    "processor_id" => 8000,
                    "submit" => "Add This File"},
                  :headers => {"content-type"=> "multipart/form-data; boundary=___23__06__24__24__12__01__38__39___"}})
    expected = [add_release, add_file]

    result = @rubyforge.requests
    result.each do |r|
      r[:form].delete "userfile"
    end

    assert_equal expected, result
  end

  def test_post_news
    @rubyforge.post_news("seattlerb", "my summary", "my news")

    util_run("/news/submit.php",
             "group_id"     => 42,
             "post_changes" => "y",
             "details"      => "my news",
             "summary"      => "my summary",
             "submit"       => "Submit")
  end

  def test_add_release_package_name
    @rubyforge.add_release(42, "woot_pkg", '1.2.3', __FILE__)
    util_add_release
  end

  def test_add_release_undefined_package_name
    assert_raise RuntimeError do
      @rubyforge.add_release(42, "blah", '1.2.3', __FILE__)
    end
  end

  def test_add_release_group_name
    @rubyforge.add_release("seattlerb", 666, '1.2.3', __FILE__)
    util_add_release
  end

  def test_add_release_undefined_group_name
    assert_raise RuntimeError do
      @rubyforge.add_release("blah", 666, '1.2.3', __FILE__)
    end
  end

  def test_lookup_id
    assert_equal 43, @rubyforge.lookup("package", 43)
  end

  def test_lookup_string_number
    assert_raise RuntimeError do
      @rubyforge.lookup("package", "43")
    end
  end

  def test_lookup_name
    @rubyforge.autoconfig["package_ids"]["ringy_dingy"] = 314
    assert_equal 314, @rubyforge.lookup("package", "ringy_dingy")
  end

  def test_lookup_undefined
    assert_raise RuntimeError do
      @rubyforge.lookup("package", "blah")
    end
  end

  def test_run
    util_new FakeRubyForge
    result = @rubyforge.add_release(42, 666, '1.2.3', __FILE__)

    assert_equal 6948, result
    extheader = {"content-type"=> "multipart/form-data; boundary=___00__03__03__39__09__19__21__36___"}

    form = {
      "processor_id" => 8000,
      "submit" => "Release File",
      "preformatted" => 0,
      "release_changes" => nil,
      "type_id" => 9999,
      "group_id" => 42,
      "release_name" => "1.2.3",
      "release_notes" => nil,
      "package_id" => 666,
      "release_date" => "today"
    }

    client = @rubyforge.client
#    assert client.form.delete("userfile")

#     assert_equal 'http://rubyforge.org/frs/admin/qrs.php', client.url.to_s
#     assert_equal form, client.form
#     assert_equal extheader, client.headers
  end

  def test_scrape_project
    orig_stdout = $stdout
    orig_stderr = $stderr
    $stdout = StringIO.new
    $stderr = StringIO.new
    util_new RubyForge # TODO: switch to Fake
    @rubyforge.autoconfig.each { |k,v| v.clear }

    URI::HTTP.data << "<a href='/tracker/?group_id=1513'>Tracker</a>"
    URI::HTTP.data << <<-EOF
<h3>ar_mailer
<a href="/frs/monitor.php?filemodule_id=4566&group_id=1513&start=1">
<a href="shownotes.php?release_id=13368">1.3.1</a>
<a href="shownotes.php?release_id=12185">1.2.0</a></strong>
    EOF

    #    @rubyforge.scrape << < <-EOF
    URI::HTTP.data << <<-EOF
<select name="processor_id">
<option value="100">Must Choose One</option>
<option value="1000">i386</option>
<option value="1001">i387</option>
</select>
    EOF

    @rubyforge.scrape_project('my_project')

    expected = {
      "group_ids" => { "my_project" => 1513 },
      "package_ids" => { "ar_mailer" => 4566 },
      "processor_ids" => { "i386" => 1000, "i387" => 1001 },
      "release_ids" => {
        "ar_mailer" => { "1.2.0" => 12185, "1.3.1" => 13368 }
      },
      "type_ids" => {},
    }

    assert_equal expected, @rubyforge.autoconfig
  ensure
    $stdout = orig_stdout
    $stderr = orig_stderr
  end

  def util_new(klass)
    @cookie_path = File.join(Dir.tmpdir, "cookie.#{$$}.dat")
    File.open(@cookie_path, 'w') {} # touch

    user_data = {
      "uri"        => "http://rubyforge.org",
      "is_private" => false,
      "cookie_jar" => @cookie_path,
      "username"   => "username",
      "password"   => "password"
    }

    auto_data = {
      "group_ids" => {},
      "package_ids" => {},
      "release_ids" => Hash.new { |h,k| h[k] = {} },
      "type_ids" => {},
      "processor_ids" => {"Any"=>8000},
    }

    @rubyforge = klass.new user_data, auto_data

    @rubyforge.client = RubyForge::FakeClient.new

    @rubyforge.userconfig["release_date"] = "today"
    @rubyforge.autoconfig["type_ids"][".rb"] = 9999
    @rubyforge.autoconfig["group_ids"]["seattlerb"] = 42
    @rubyforge.autoconfig["package_ids"]["woot_pkg"] = 666
  end

  def util_run(page, form={}, extheader={})
    form_result = @rubyforge.form
    assert form_result.delete("userfile") unless extheader.empty?

    assert_equal page, @rubyforge.page.to_s
    assert_equal form, form_result
    assert_equal extheader, @rubyforge.extheader
  end

  def util_add_release
    util_run('/frs/admin/qrs.php',
             {"processor_id" => 8000,
               "submit" => "Release File",
               "preformatted" => 0,
               "release_changes" => nil,
               "type_id" => 9999,
               "group_id" => 42,
               "release_name" => "1.2.3",
               "release_notes" => nil,
               "package_id" => 666,
               "release_date" => "today"},
             {"content-type"=> "multipart/form-data; boundary=___00__03__03__39__09__19__21__36___"})
  end

  def util_delete_package
    util_run('/frs/admin/index.php',
             "submit" => "Delete",
             "really_sure" => "1",
             "group_id" => 42,
             "func" => "delete_package",
             "package_id" => 666,
             "sure" => "1")
  end
end


test/test_rubyforge_client.rb
require 'test/unit' unless defined? $ZENTEST and $ZENTEST
require 'rubyforge'

class TestRubyForgeClient < Test::Unit::TestCase
  def setup
    RubyForge::Client.const_set(:Net, Module.new)
    RubyForge::Client::Net.const_set(:HTTP, Class.new(DelegateClass(::Net::HTTP)))

    RubyForge::Client::Net.class_eval do
      def self.const_missing(sym)
        ::Net.const_get(sym)
      end
    end

    RubyForge::Client::Net::HTTP.class_eval do
      def initialize(*args)
        @http = ::Net::HTTP.new(*args)
        super(@http)
      end

      class << self; attr_accessor :t_request, :t_data; end
      def request(request, data)
        self.class.instance_variable_set(:@t_request, request)
        self.class.instance_variable_set(:@t_data, data)
        response = Net::HTTPOK.new('1.1', 200, '')
        class << response
          def read_body; ''; end
        end
        return response
      end

      def self.const_missing(sym)
        ::Net::HTTP.const_get(sym)
      end
    end

    @client = RubyForge::Client.new
  end

  def teardown
    RubyForge::Client.class_eval { remove_const(:Net) }
  end

  def test_post_with_params
    @client.post_content('http://example.com', { :f => 'adsf'})
    assert_equal('f=adsf', RubyForge::Client::Net::HTTP.t_data)

    @client.post_content('http://example.com', { :a => 'b', :c => 'd' })
    assert_equal('a=b&c=d', RubyForge::Client::Net::HTTP.t_data)
  end

  def test_multipart_post_one_param
    random = Array::new(8){ "%2.2d" % rand(42) }.join('__')
    boundary = "multipart/form-data; boundary=___#{ random }___"

    request = <<-END
--___#{random}___\r
Content-Disposition: form-data; name="a"\r\n\r
b\r
--___#{random}___--\r
END

    @client.post_content( 'http://example.com',
                          { :a => 'b' },
                          { 'content-type' => boundary }
                        )
    assert_equal(request, RubyForge::Client::Net::HTTP.t_data)
  end

  def test_multipart_post_two_params
    random = Array::new(8){ "%2.2d" % rand(42) }.join('__')
    boundary = "multipart/form-data; boundary=___#{ random }___"

    request = <<-END
--___#{random}___\r
Content-Disposition: form-data; name="a"\r\n\r
b\r
--___#{random}___\r
Content-Disposition: form-data; name="c"\r\n\r
d\r
--___#{random}___--\r
END

    @client.post_content( 'http://example.com',
                          { :a => 'b', :c => 'd' },
                          { 'content-type' => boundary }
                        )
    assert_equal(request, RubyForge::Client::Net::HTTP.t_data)
  end

  def test_multipart_io
    random = Array::new(8){ "%2.2d" % rand(42) }.join('__')
    boundary = "multipart/form-data; boundary=___#{ random }___"

    file_contents = 'blah blah blah'
    file = StringIO.new(file_contents)
    class << file
      def path
        '/one/two/three.rb'
      end
    end

    request = <<-END
--___#{random}___\r
Content-Disposition: form-data; name="userfile"; filename="three.rb"\r
Content-Transfer-Encoding: binary\r
Content-Type: text/plain\r\n\r
#{file_contents}\r
--___#{random}___--\r
END

    @client.post_content( 'http://example.com',
                          { :userfile => file },
                          { 'content-type' => boundary }
                        )
    assert_equal(request, RubyForge::Client::Net::HTTP.t_data)
  end
end


lib/rubyforge/client.rb
require 'webrick/cookie'
require 'net/http'
require 'net/https'
require 'rubyforge/cookie_manager'

# clean up warnings caused by web servers that send down 2 digit years
class Time
  CENTURY = Time.now.year / 100 * 100

  class << self
    alias :old_utc :utc
    def utc(*args)
      args[0] += CENTURY if args[0] < 100 if args.length < 10
      old_utc(*args)
    end
  end
end unless Time.respond_to? :old_utc

# clean up "using default DH parameters" warning for https
class Net::HTTP
  alias :old_use_ssl= :use_ssl=
  def use_ssl= flag
    self.old_use_ssl = flag
    @ssl_context.tmp_dh_callback = proc {} if @ssl_context
  end
end unless Net::HTTP.public_instance_methods.include? "old_use_ssl="

class RubyForge
  class Client
    attr_accessor :debug_dev, :ssl_verify_mode, :agent_class

    def initialize(proxy = nil)
      @debug_dev       = nil
      @ssl_verify_mode = OpenSSL::SSL::VERIFY_NONE
      @cookie_manager  = CookieManager.new
      if proxy
        begin
          proxy_uri = URI.parse(proxy)
          @agent_class = Net::HTTP::Proxy(proxy_uri.host,proxy_uri.port)
        rescue URI::InvalidURIError
        end
      end
      @agent_class ||= Net::HTTP
    end

    def cookie_store
      @cookie_manager
    end

    def cookie_store=(path)
      @cookie_manager = CookieManager.load(path)
    end

    def post_content(uri, form = {}, headers = {})
      uri = URI.parse(uri) unless uri.is_a?(URI)
      request = agent_class::Post.new(uri.request_uri)
      execute(request, uri, form, headers)
    end

    def get_content(uri, query = {}, headers = {})
      uri = URI.parse(uri) unless uri.is_a?(URI)
      request = agent_class::Get.new(uri.request_uri)
      execute(request, uri, query, headers)
    end

    def execute(request, uri, parameters = {}, headers = {})
      {
        'content-type' => 'application/x-www-form-urlencoded'
      }.merge(headers).each { |k,v| request[k] = v }

      @cookie_manager[uri].each { |k,v|
        request['Cookie'] = v.to_s
      } if @cookie_manager[uri]

      http = agent_class.new( uri.host, uri.port )

      if uri.scheme == 'https'
        http.use_ssl      = true
        http.verify_mode  = OpenSSL::SSL::VERIFY_NONE
      end

      request_data = case request['Content-Type']
                     when /boundary=(.*)$/
                       boundary_data_for($1, parameters)
                     else
                       query_string_for(parameters)
                     end
      request['Content-Length'] = request_data.length.to_s

      response = http.request(request, request_data)
      (response.get_fields('Set-Cookie') || []).each do |raw_cookie|
        WEBrick::Cookie.parse_set_cookies(raw_cookie).each { |baked_cookie|
          baked_cookie.domain ||= url.host
          baked_cookie.path   ||= url.path
          @cookie_manager.add(uri, baked_cookie)
        }
      end

      return response.body if response.class <= Net::HTTPSuccess

      if response.class <= Net::HTTPRedirection
        location = response['Location']
        unless location =~ /^http/
          location = "#{uri.scheme}://#{uri.host}#{location}"
        end
        uri = URI.parse(location)

        execute(agent_class::Get.new(uri.request_uri), uri)
      end
    end

    def boundary_data_for(boundary, parameters)
      parameters.sort_by {|k,v| k.to_s }.map { |k,v|
        parameter = "--#{boundary}\r\nContent-Disposition: form-data; name=\"" +
            WEBrick::HTTPUtils.escape_form(k.to_s) + "\""

        if v.respond_to? :path
          parameter += "; filename=\"#{File.basename(v.path)}\"\r\n"
          parameter += "Content-Transfer-Encoding: binary\r\n"
          parameter += "Content-Type: text/plain"
        end
        parameter += "\r\n\r\n"

        if v.respond_to? :path
          parameter += v.read
        else
          parameter += v.to_s
        end

        parameter
      }.join("\r\n") + "\r\n--#{boundary}--\r\n"
    end

    def query_string_for(parameters)
      parameters.sort_by {|k,v| k.to_s }.map { |k,v|
        k && [  WEBrick::HTTPUtils.escape_form(k.to_s),
                WEBrick::HTTPUtils.escape_form(v.to_s) ].join('=')
      }.compact.join('&')
    end

    def save_cookie_store
      @cookie_manager.save!
    end
  end
end


lib/rubyforge/cookie_manager.rb
class RubyForge
  class CookieManager
    class << self
      def load(path)
        cm = YAML.load_file(path) || CookieManager.new(path)
        cm = CookieManager.new(path) unless cm.is_a?(CookieManager)
        cm.clean_stale_cookies
      end
    end

    attr_accessor :cookies_file
    def initialize(cookies_file = nil)
      @jar = Hash.new { |hash,domain_name|
        hash[domain_name.downcase] = {}
      }
      @cookies_file = cookies_file
    end

    def [](uri)
      # FIXME we need to do more matching on hostname....  This is not
      # bulletproof
      @jar[uri.host.downcase]
    end

    def empty?
      @jar.empty? || @jar.all? { |k,v| v.empty? }
    end

    def save!
      File.open(@cookies_file, 'wb') { |f|
        f.write(YAML.dump(self))
      }
    end

    def add(uri, cookie)
      no_dot_domain = cookie.domain.gsub(/^\./, '')
      return unless uri.host =~ /#{no_dot_domain}$/i
      @jar[no_dot_domain][cookie.name] = cookie
      clean_stale_cookies
    end

    def clean_stale_cookies
      @jar.each do |domain, cookies|
        cookies.each do |name, cookie|
          cookies.delete(name) if cookie.expires < Time.now
        end
      end
      self
    end
  end
end


lib/rubyforge.rb
#! /usr/bin/env ruby -w

require 'json'
require 'enumerator'
require 'fileutils'
require 'yaml'
require 'open-uri'
require 'rubyforge/client'

$TESTING = false unless defined? $TESTING

class RubyForge

  # :stopdoc:
  VERSION     = '2.0.2'
  HOME        = ENV["HOME"] || ENV["HOMEPATH"] || File::expand_path("~")
  RUBYFORGE_D = File::join HOME, ".rubyforge"
  CONFIG_F    = File::join RUBYFORGE_D, "user-config.yml"

  # We must use __FILE__ instead of DATA because this is now a library
  # and DATA is relative to $0, not __FILE__.
  config = File.read(__FILE__).split(/__END__/).last.gsub(/#\{(.*)\}/) {eval $1}
  CONFIG = YAML.load(config)
  # :startdoc:

  # TODO: add an autoconfig method that is self-repairing, removing key checks 
  attr_reader :userconfig, :autoconfig

  def initialize(userconfig=nil, autoconfig=nil, opts=nil)
    # def initialize(userconfig=CONFIG_F, opts={})
    @userconfig, @autoconfig = userconfig, autoconfig

    @autoconfig ||= CONFIG["rubyforge"].dup
    @userconfig.merge! opts if opts

    @client = nil
    @uri = nil
  end
  
  # These are no-ops now, but we'll keep them here for backwards compatibility
  def login ; end
  def logout ; end

  def configure opts = {}
    user_path        = CONFIG_F
    dir, file        = File.split(user_path)

    @userconfig      = if test(?e, user_path) then
                         YAML.load_file(user_path)
                       else
                         CONFIG
                       end.merge(opts)
    @autoconfig_path = File.join(dir, file.sub(/^user/, 'auto'))
    @autoconfig      = if test(?e, @autoconfig_path) then
                         YAML.load_file(@autoconfig_path)
                       else
                         CONFIG["rubyforge"].dup
                       end
    @autoconfig["type_ids"] = CONFIG['rubyforge']['type_ids'].dup

    raise "no <username>"   unless @userconfig["username"]
    raise "no <password>"   unless @userconfig["password"]

    self
  end

  def force
    @userconfig['force']
  end

  def uri
    uri = @userconfig['uri']
    abort "Using new REST api, but uri isn't api.rubyforge.org.
run `rubyforge setup` and fix please" if
      uri =~ /rubyforge.org/ and uri !~ /api.rubyforge.org/

    @uri ||= URI.parse uri
  end

  def setup
    FileUtils::mkdir_p RUBYFORGE_D, :mode => 0700 unless test ?d, RUBYFORGE_D
    test ?e, CONFIG_F and FileUtils::mv CONFIG_F, "#{CONFIG_F}.bak"
    config = CONFIG.dup
    config.delete "rubyforge"

    open(CONFIG_F, "w") { |f|
      f.write YAML.dump(config)
    }
    edit = (ENV["EDITOR"] || ENV["EDIT"] || "vi") + " '#{CONFIG_F}'"
    system edit or puts "edit '#{CONFIG_F}'"
  end

  def save_autoconfig
    File.open(@autoconfig_path, "w") do |file|
      YAML.dump @autoconfig, file
    end
  end

  def scrape_config
    username = @userconfig['username']

    %w(group package processor release).each do |type|
      @autoconfig["#{type}_ids"].clear if @autoconfig["#{type}_ids"]
    end
  
    json = get_via_rest_api "/users/#{username}/groups.js"

    projects = json.collect {|group| group['group']['unix_group_name'] }
    puts "Fetching #{projects.size} projects"
    projects.each do |project|
      scrape_project(project)
    end
  end
  
  def get_via_rest_api(path)
    url = "#{self.uri}#{path}"
    puts "Hitting REST API: #{url}" if $DEBUG
    JSON.parse(client.get_content(url, {}, {}, @userconfig))
  end

  def scrape_project(project)
    data = {
      "group_ids"     => {},
      "package_ids"   => {},
      "processor_ids" => Hash.new { |h,k| h[k] = {} },
      "release_ids"   => Hash.new { |h,k| h[k] = {} },
    }

    unless data["group_ids"].has_key? project then
      json = get_via_rest_api "/groups/#{project}.js"
      group_id = json["group"]["group_id"].to_i
      data["group_ids"][project] = group_id
    end

    # Get project's packages 
    json = get_via_rest_api "/groups/#{project}/packages.js"
    json.each do |package|
      data["package_ids"][package["package"]["name"]] = package["package"]["package_id"]
      # Get releases for this package
      json = get_via_rest_api "/packages/#{package["package"]["package_id"]}/releases.js"
      json.each do |release|
        data["release_ids"][package["package"]["name"]][release["name"]] = release["release_id"]
      end
    end

    # Get processor ids
    if @autoconfig['processor_ids'].nil? || @autoconfig['processor_ids'].empty?
      puts "Fetching processor ids" if $DEBUG
      json = get_via_rest_api "/processors.js"
      json.each do |processor|
        data["processor_ids"][processor["processor"]["name"]] = processor["processor"]["processor_id"]
      end
    end

    data.each do |key, val|
      @autoconfig[key] ||= {}
      @autoconfig[key].merge! val
    end

    save_autoconfig
  end

  def create_package(group_id, package_name)
    page = "/groups/#{group_id}/packages"

    group_id = lookup "group", group_id
    is_private = @userconfig["is_private"]
    is_public = is_private ? 0 : 1

    form = {
      "package[name]" => package_name,
      "package[is_public]"    => is_public
    }

    run page, form

    group_name = @autoconfig["group_ids"].invert[group_id]
    scrape_project(group_name)
  end

  ##
  # Posts news item to +group_id+ (can be name) with +subject+ and +body+

  def post_news(group_id, subject, body)
    # TODO - what was the post_changes parameter for?
    form = {
      "news_byte[summary]"      => subject,
      "news_byte[details]"      => body
    }
    group_id = lookup "group", group_id
    url = "/groups/#{group_id}/news_bytes"
    run url, form
  end

  def delete_package(group_id, package_id)
    group_id = lookup "group", group_id
    package_id = lookup "package", package_id
    package_name = @autoconfig["package_ids"].invert[package_id]
    @autoconfig["package_ids"].delete package_name
    @autoconfig["release_ids"].delete package_name
    save_autoconfig
    url = "/packages/#{package_id}"
    run url, {"_method" => "delete"}
  end

  def add_release(group_id, package_id, release_name, *files)
    group_id        = lookup "group", group_id
    package_id      = lookup "package", package_id
    release_date    = @userconfig["release_date"]
    release_notes   = @userconfig["release_notes"]
    release_changes = @userconfig["release_changes"]
    preformatted    = @userconfig["preformatted"]
    release_date ||= Time.now.strftime("%Y-%m-%d %H:%M")
    release_notes = IO::read(release_notes) if
      test(?e, release_notes) if release_notes
    release_changes = IO::read(release_changes) if
      test(?e, release_changes) if release_changes
    preformatted = preformatted ? 1 : 0

    form = {
        "release[name]"    => release_name,
        "release[release_date]"     => release_date,
        "release[notes]"   => release_notes,
        "release[changes]" => release_changes,
        "release[preformatted]"    => preformatted,
    }

    url = "/packages/#{package_id}/releases"
    json = run url, form
    
    release_id = JSON.parse(json)["release_id"].to_i rescue nil
    unless release_id then
      puts json if $DEBUG
      raise "Couldn't get release_id, upload failed?"
    end

    # FIXME
    #raise "Invalid package_id #{package_id}" if html[/Invalid package_id/]
    #raise "You have already released this version." if html[/That filename already exists in this project/]

    files.each do |file|
      add_file(group_id, package_id, release_id, file)
    end

    package_name = @autoconfig["package_ids"].invert[package_id]
    raise "unknown package name for #{package_id}" if package_name.nil?
    @autoconfig["release_ids"][package_name] ||= {}
    @autoconfig["release_ids"][package_name][release_name] = release_id
    save_autoconfig

    release_id
  end

  ##
  # add a file to an existing release under the specified group_id,
  # package_id, and release_id
  #
  # example :
  #   add_file("codeforpeople", "traits", "0.8.0", "traits-0.8.0.gem")
  #   add_file("codeforpeople", "traits", "0.8.0", "traits-0.8.0.tgz")
  #   add_file(1024, 1242, "0.8.0", "traits-0.8.0.gem")

  def add_file(group_name, package_name, release_name, userfile)
    type_id      = @userconfig["type_id"]
    group_id     = lookup "group", group_name
    package_id   = lookup "package", package_name
    release_id   = (Integer === release_name) ? release_name : lookup("release", package_name)[release_name]
    url = "/releases/#{release_id}/files.js"

    userfile = open userfile, 'rb'

    type_id ||= userfile.path[%r|\.[^\./]+$|]
    type_id = (lookup "type", type_id rescue lookup "type", ".oth")

    processor_id = @userconfig["processor_id"]
    processor_id ||= "Any"
    processor_id = lookup "processor", processor_id

    form = {
      "file[filename]"      => File.basename(userfile.path),
      "file[processor_id]"  => processor_id,
      "file[type_id]"       => type_id,
      "contents"            => userfile.read
    }

    run url, form
  end

  def client
    return @client if @client

    @client = RubyForge::Client::new ENV["HTTP_PROXY"]
    @client.debug_dev = STDERR if ENV["RUBYFORGE_DEBUG"] || ENV["DEBUG"] || $DEBUG

    @client
  end

  def run(page, form, extheader={}) # :nodoc:
    uri = self.uri + page
    puts "client.post_content #{uri.inspect}, #{form.inspect}, #{extheader.inspect}" if $DEBUG
    response = client.post_content uri, form, extheader, @userconfig
    puts response if $DEBUG
    response
  end

  def lookup(type, val) # :nodoc:
    unless Fixnum === val then
      key = val.to_s
      val = @autoconfig["#{type}_ids"][key]
      raise "no <#{type}_id> configured for <#{ key }>" unless val
    end
    val
  end
end

__END__
#
# base rubyforge uri - store in #{ CONFIG_F }
#
  uri        : http://api.rubyforge.org
#
# this must be your username
#
  username   : tom
#
# this must be your password
#
  password   : password
#
# defaults for some values
#
  is_private : false
# AUTOCONFIG:
  rubyforge :
  #
  # map your group names to their rubyforge ids
  #
    group_ids :
      codeforpeople : 1024
      support : 5
  #
  # map your package names to their rubyforge ids
  #
    package_ids :
      traits : 1241
  #
  # map your package names to their rubyforge ids
  #
    release_ids :
      traits :
        1.2.3 : 666
  #
  # mapping file exts to rubyforge ids
  #
    type_ids :
      .deb         : 1000
      .rpm         : 2000
      .zip         : 3000
      .bz2         : 3100
      .gz          : 3110
      .src.zip     : 5000
      .src.bz2     : 5010
      .src.tar.bz2 : 5010
      .src.gz      : 5020
      .src.tar.gz  : 5020
      .src.rpm     : 5100
      .src         : 5900
      .jpg         : 8000
      .txt         : 8100
      .text        : 8100
      .htm         : 8200
      .html        : 8200
      .pdf         : 8300
      .oth         : 9999
      .ebuild      : 1300
      .exe         : 1100
      .dmg         : 1200
      .tar.gz      : 5000
      .tgz         : 5000
      .gem         : 1400
      .pgp         : 8150
      .sig         : 8150
      .pem         : 1500

  #
  # map processor names to rubyforge ids
  #
    processor_ids :
      Other      : 9999
